### PR TITLE
chore(lib/grandpa): `lib/grandpa/models` package

### PR DESCRIPTION
**Pull request by qdm12**
## Changes

- Dumb dumb moving of files from `lib/grandpa` to `lib/grandpa/models`
- Minimal diffs, only affects `lib/grandpa`

Why: because adding tests in `lib/grandpa` drove me mad having to fiddle with all the implementation details which can be abstracted with interfaces and (internal) sub-packages, since I don't want to pollute the grandpa library Go API even further with 'internally-used-only' interfaces. This is the nth time this triggers me as we can't break lib/grandpa in subpackages, and it makes development harder.

Also let me know if you prefer another name than `models` for the bunch of messages and other structs. Maybe we can name it `messages` :thinking: 

Other PRs should be made subsequently (to alleviate review load, this one is just moving):

1. Move `State` from `lib/grandpa/models` to `lib/grandpa` as `state` (5 minutes job)
1. Re-arrange grandpa models files a bit (5 minutes job) - see Eclesio's comment below!
1. Move models only used internally by lib/grandpa to `internal/grandpa/models` (20min job)

## Tests

```sh
go test -tags integration github.com/ChainSafe/gossamer/lib/grandpa/...
```

## Issues

Unblocks #2506

### CODE REVIEW

**Comment by EclesioMeloJunior** [comment]
I would suggest keeping the messages variadic data types in the messages pkg each in a file. I would suggest renaming the `types.go` file to `round.go` as I think the contents are not related to ˜types˜. This will add more deltas however, I think will help us in the future to understand better the code and shrink even more the grandpa package.
```
- lib/grandpa
  |- messages/
    |- grandpa_message.go
    |- vote_message.go
    |- neighbour_message.go
    |- commit_message.go
    |- catch_up_request
    ...
  |- types.go -> round.go
```

**Review by EclesioMeloJunior — CHANGES_REQUESTED** [review]

File: lib/grandpa/models/errors.go
Comment (qdm12) [reply lib/grandpa/models/errors.go]: Actually to reduce package dependency (and avoid cycles), I created a `lib/grandpa/errors` package with `errors.go` listing those 2 errors, which can be returned from some functions both in `lib/grandpa` and in `lib/grandpa/models`. I also removed them from `lib/grandpa`, to avoid having errors defined twice (and since `lib/gandpa` imports `lib/grandpa/models`, `lib/grandpa/models` cannot import errors from its importer `lib/grandpa`).

**Comment by qdm12** [comment]
@EclesioMeloJunior I'll keep it as `models` instead of:

- `messages` in this PR, since it contains messages and some other types. I'll create a `messages` package in another PR.
- `types` since it then conflicts with `dot/types` imports, so I thought `models` just works fine in this case 😄

**Comment by EclesioMeloJunior** [comment]
@qdm12
> * `messages` in this PR, since it contains messages and some other types. I'll create a `messages` package in another PR.

I don't see any problem with having a massage package that contains related types and all the files in the `message` package will have very close types. On second thought, I believe that the `messages` package should be in `lib/` instead of `/lib/grandpa` package because we have other messages e.g `BlockAnnounce` not related to grandpa that can fit in the package

```
- lib/
  |- messages/
    |- grandpa_message.go
    |- vote_message.go
    |- neighbour_message.go
    |- commit_message.go
    |- catch_up_request
    ...
  |- grandpa/
     |- ...
```

I notice that none of the messages have strong dependency with the `grandpa` package, most of the dependencies in the `message` is the package `lib/grandpa/errors` which can be moved to a file inside the package messages.

> * `types` since it then conflicts with `dot/types` imports, so I thought `models` just works fine in this case 😄

I did not suggest creating a `types` package but renaming the `types.go` file to `round.go` and moving it from `messages`  to `lib/grandpa` folder. 

```
- lib/
  |- grandpa/
     |- types.go -> round.go
```

I'm suggesting a reorganization to avoid the generic name `models`, for example, in this PR you introduce the `models` but you import it in the `lib/grandpa/grandpa.go` as:

```
import (
	grandpa "github.com/ChainSafe/gossamer/lib/grandpa/models"
)
```

and IMO this is some kind of red flag that we should avoid.

**Comment by qdm12** [comment]
> I don't see any problem with having a massage package that contains related types and all the files in the `message` package will have very close types. On second thought, I believe that the `messages` package should be in `lib/` instead of `/lib/grandpa` package because we have other messages e.g `BlockAnnounce` not related to grandpa that can fit in the package

Completely agreed, but in the next PR (right after this gets merged) to keep track of file movements, notably:

- [lib/grandpa/message.go → lib/grandpa/models/message.go](https://github.com/ChainSafe/gossamer/pull/2575/files#diff-84486edbf6ccfe9c259c19832e618ab1831b848c8f2019109fe67ed21c652795)
- [lib/grandpa/types.go → lib/grandpa/models/types.go](https://github.com/ChainSafe/gossamer/pull/2575/files#diff-675f35ea8ffd4d2309bbc8498cac44679485bf41c76a631c0aaa8c2dc1c7dbdf)
- [lib/grandpa/types_test.go → lib/grandpa/models/types_test.go](https://github.com/ChainSafe/gossamer/pull/2575/files#diff-4cf51db473b08b90081058978dbf4d824d8868ede2458814092d04914c3180db)

> most of the dependencies in the message is the package lib/grandpa/errors which can be moved to a file inside the package messages.

Totally, sorry i lost myself it seems. Both errors are moved to `lib/grandpa/models/errors.go` for now. They will get moved to `lib/messages` in an upcoming PR.


> I'm suggesting a reorganization to avoid the generic name `models`, for example, in this PR you introduce the `models` but you import it in the `lib/grandpa/grandpa.go` as:
> 
> ```
> import (
> 	grandpa "github.com/ChainSafe/gossamer/lib/grandpa/models"
> )
> ```
> 
> and IMO this is some kind of red flag that we should avoid.

As mentioned, I totally agree, but in the next PR to keep track of file movements (and to not have a 10K lines PR as well).
Also it is not imported like such in `lib/grandpa/grandpa.go`. It is however imported in other packages (`dot/digest` and `dot/rpc/subscription`) as `grandpa "github.com/ChainSafe/gossamer/lib/grandpa/models"` to keep deltas to a minimum in this PR. But yeah it should be changed to `lib/messages` in an upcoming PR, although those references will probably still use the `grandpa` alias in the import to keep deltas low, until the last PRs to finally put a nail into all this.

**Review by EclesioMeloJunior — APPROVED** [review]
Sorry to be pedantic on this subject but I got the point 😅

**Comment by qdm12** [comment]
> Sorry to be pedantic on this subject

No apologizing needed at all. I'm the pedanticest (time to create that word) with all my nits, so I can't complain 🤣 
And also good comments that got used here (errors in the models package) and other that will be used for the next PRs! 🚀

**Review by timwu20 — CHANGES_REQUESTED** [review]
Regarding the `models` package, from the [go blog](https://go.dev/blog/package-names):

> Don’t use a single package for all your APIs. Many well-intentioned programmers put all the interfaces exposed by their program into a single package named api, types, or interfaces, thinking it makes it easier to find the entry points to their code base. This is a mistake. Such packages suffer from the same problems as those named util or common, growing without bound, providing no guidance to users, accumulating dependencies, and colliding with other imports. Break them up, perhaps using directories to separate public packages from implementation.

I don't see why we need to move grandpa at this point into subpackages.  

> Why: because adding tests in lib/grandpa drove me mad having to fiddle with all the implementation details which can be abstracted with interfaces and (internal) sub-packages, since I don't want to pollute the grandpa library Go API even further with 'internally-used-only' interfaces. This is the nth time this triggers me as we can't break lib/grandpa in subpackages, and it makes development harder.

What's wrong with internal only interfaces? There are a number of reasons why using internal interfaces is beneficial.

I don't see the reason why this PR needed to be opened at this time.  We have resources currently working on grandpa, and this will slow them down if this gets merged in.  We can address larger refactors later as a group.  This type of task should be more exploratory, we should have discussed it as a group first.

**Comment by qdm12** [comment]
> Don’t use a single package for all your APIs. Many well-intentioned programmers put all the interfaces exposed by their program into a single package named api, types, or interfaces

> I don't see why we need to move grandpa at this point into subpackages.

Not applicable here, we use those messages structures in grandpa and other packages, so from a production-code standpoint, it does make sense to have it as a separate package. These messages are not necessarily tied to grandpa. From a testing standpoint, not having subpackages (i.e. for the vote messages tracker) makes it hell to write (deep) tests for the grandpa package.

> This type of task should be more exploratory, we should have discussed it as a group first.

I have discussed this in standup in ~February, we decided to not do it, but when adding tests for code surrounding the trackers in April/May, I really got blocked by this again, and I did mention I would do it in standup, and opened this first PR.

I could again forget about it and forget about finishing those tests, and sort of know the grandpa package is in a hard-to-test state, but I would rather slowly improve it.

> What's wrong with internal only interfaces? There are a number of reasons why using internal interfaces is beneficial.

Well, in my opinionated opinion (which I still quite believe in so far), an internal interface is a sign that either:

- it should not be an interface; it should be a concrete implementation, and tests should not mock it; or
- the implementation should be in a separate package, and injected to the package as an exported interface via an exported constructor. That motivates for more dependency injection, which helps in many ways

I wrote a whole bunch of go code without ever using private interfaces, and only doing exported interfaces + concrete unexported structs, and that served me well so far.

> We have resources currently working on grandpa, and this will slow them down if this gets merged in.

This PR tracks the file movements, so a git rebase would be rather trivial for other developers.

**Comment by timwu20** [comment]
Let's revisit this after the grandpa refactors and after the implementation of polite grandpa #2505 .

**Pull request closed without merging**
